src/main/java/org/elasticsearch/rest/BaseRestHandler.java

```diff
@@ -66,12 +66,7 @@ public final void handleRequest(RestRequest request, RestChannel channel) throws
         }
 
         private static void copyHeadersAndContext(ActionRequest actionRequest, RestRequest restRequest, Set<String> headers) {
-            for (String usefulHeader : headers) {
-                String headerValue = restRequest.header(usefulHeader);
-                if (headerValue != null) {
-                    actionRequest.putHeader(usefulHeader, headerValue);
-                }
-            }
+            actionRequest.copyHeadersFrom(actionRequest);
```

Comment:
I had to restore the old behaviour, as only specified/registered HTTP headers are included in the real request... added a test as well